### ISSUE TITLE
Unnest causes multiple invocations of map_batches function

### ISSUE THREAD

**Issue by JeppeLS** [pola-rs/polars#20260]
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl


def f(s: pl.Series) -> pl.Series:
    print("Invoked")
    return pl.Series(
        "x",
        [
            {"y": 1.0, "z": 2.0},
        ],
    )


df = pl.DataFrame({"x": [1.0]})

df = df.with_columns(
    pl.col("x")
    .map_batches(f, return_dtype=pl.Struct({"y": pl.Float64, "z": pl.Float64}))
    .struct.unnest()
)
```

### Log output

```shell
Invoked
Invoked
```

### Issue description

It seems like the UDF given to `map_batches` is invoked an additional time when it is combined with `.struct.unnext()`. As far as I can tell it should only be invoked once. If I split it into two calls, the UDF is only invoked once i.e
```python
df = df.with_columns(
    pl.col("x").map_batches(
        f, return_dtype=pl.Struct({"y": pl.Float64, "z": pl.Float64})
    )
)
df = df.with_columns(pl.col("x").struct.unnest())
```

only prints a single `Invoked`

### Expected behavior

The UDF should be called only once so only a single `Invoked` should be printed.

### Installed versions

<details>

```
Polars:              1.17.1
Index type:          UInt32
Platform:            Linux-6.8.0-45-generic-x86_64-with-glibc2.39
Python:              3.12.7 | packaged by conda-forge | (main, Oct  4 2024, 16:05:46) [GCC 13.3.0]
LTS CPU:             False

----Optional dependencies----
adbc_driver_manager  1.3.0
altair               5.5.0
boto3                1.35.65
cloudpickle          <not installed>
connectorx           0.4.0
deltalake            <not installed>
fastexcel            <not installed>
fsspec               2024.10.0
gevent               <not installed>
google.auth          <not installed>
great_tables         <not installed>
matplotlib           3.9.2
nest_asyncio         1.6.0
numpy                2.1.3
openpyxl             <not installed>
pandas               2.2.3
pyarrow              18.0.0
pydantic             <not installed>
pyiceberg            <not installed>
sqlalchemy           <not installed>
torch                2.5.1+cu124
xlsx2csv             <not installed>
xlsxwriter           <not installed>
```

</details>

**Comment by ritchie46** [comment]
That's expected. Unnest who expand the expression and we don't cse python functions.

**Comment by talawahtech** [comment]
I just gotten bitten by this as well. It would be great to get UDF support for common sub-plan elimination. In my case the simple fix of throwing another `select/with_columns` in the mix doesn't work because I am using `map_batches` in a helper function that just uses Expressions and doesn't have the DataFrame context.

Here is a dumbed down example. In reality I am getting the stats from an HDR histogram encoded into a binary blob:

```
ldf = pl.DataFrame({ "data": [i for i in range(100)] }).lazy()

def udf_stats(data: list[int]) -> dict[str, int]:
    return { "min": min(data), "max": max(data), "sum": sum(data), "count": len(data) }

def get_stats_cols(data_col_name: str = "data", use_udf_stats = False) -> pl.Expr | list[pl.Expr]:
    data_col = pl.col(data_col_name)

    if use_udf_stats:
        return_dtype = pl.Struct({ "min": pl.Int64, "max": pl.Int64, "sum": pl.Int64, "count": pl.Int64 })
        return data_col.map_batches(udf_stats, return_dtype).struct.unnest()
    else:
        return [
            data_col.min().alias("min"),
            data_col.max().alias("max"),
            data_col.sum().alias("sum"),
            data_col.len().alias("count"),
        ]

stats_columns = get_stats_cols(use_udf_stats=True)
stats_df = ldf.select(stats_columns).collect()
```

For now, my workaround is to add `@lru_cache` to `udf_stats()`, but if anyone else has a better workaround until this is fixed I'm all ears.

**Comment by gab23r** [comment]
This was fixed by https://github.com/pola-rs/polars/pull/26253
